rpc/src/optimistically_confirmed_bank_tracker.rs

```diff
@@ -402,7 +402,7 @@ mod tests {
     };
 
     /// Given a bank get the parent slot chains, this include the parent slot of the oldest parent bank.
-    fn get_parent_chains(bank: &Arc<Bank>) -> Vec<Slot> {
+    fn get_parent_chains(bank: Arc<Bank>) -> Vec<Slot> {
```

Comment:
Follow-up: Don't we essentially have this fn already in `Bank::parents_inclusive`?

Even so this kinda seems like we don't need to do any cloning? Just walk the parents back until we find the oldest one, but only copy slots.

Reply:
If it is not a bug, it is very confusing.
Wouldn't `oldest_parent` always be slot `0`, except for slot `0` itself?

Why `bank.parent_slot()` is a good default for the last parent slot, if it is absent? %)
I mean, from the standpoint of which slots are rooted.
If the above assumption is correct, it will still always be only `0`.
So it is not helping in any way, except making it more confusing.

Reply:
> Wouldn't oldest_parent always be slot 0, except for slot 0 itself?

- Theoretically speaking - yes; any block should be able to walk its' ancestry back to block 0
- Functionally speaking - no; the walk is happening on Banks. At runtime, Banks are somewhat expensive from a resource standpoint so we try to be fairly aggressive in cleaning them up. Once a slot `R` has been rooted, all banks with slot `< R` are cleaned. So, in this case, if `R` was rooted, `R`'s parent will be cleaned up and `R`'s parent will be set to `None`. Check out `Bank::squash()` + `BankForks::do_set_root_return_metrics()`.

> Why bank.parent_slot() is a good default for the last parent slot, if it is absent? %)

For some bank `B` with parent `P`, `B` stashes a copy of `P`'s slot. So, even if `P` gets cleaned up, `B` can still know what slot `P` was.
https://github.com/solana-labs/solana/blob/0996b5e245d741d6ccbf3ef0e5c01bc3297b6e63/runtime/src/bank.rs#L678

In any case, I agree that this function is much more confusing than it should be. And, I think Andrew's suggestion of using `Bank::parents_inclusive()` will do the job while being much more clear.